### PR TITLE
Allow consumers of `XmlFile` to disable `fsync(2)`

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-66001](https://issues.jenkins.io/browse/JENKINS-66001). Part 1 of a 5-part series to make Pipeline's `SURVIVABLE_NONATOMIC` mode behave [as advertised in the documentation](https://www.jenkins.io/doc/book/pipeline/scaling-pipeline/#what-am-i-giving-up-with-this-durability-setting-trade-off), namely by using `AtomicFileWriter` but retaining the skipping of `fsync(2)` (parts 2 through 5 are in jenkinsci/workflow-api-plugin#163, jenkinsci/workflow-support-plugin#120, jenkinsci/workflow-cps-plugin#452, and jenkinsci/workflow-job-plugin#199, respectively). This Jenkins core PR provides a new API that is consumed by downstream PRs mentioned above.

I tested this by running a Pipeline job with 100 steps in `SURVIVABLE_NONATOMIC` mode while simultaneously attaching a remote debugger to `XmlFile` and monitoring `fsync(2)` calls with [`syncsnoop.bt`](https://github.com/iovisor/bpftrace/blob/6cb740241edddc4b88e039bfc9385e6a3bad3abb/tools/syncsnoop.bt). I confirmed that `AtomicFileWriter` was being used in the Java debugger, and I confirmed that `fsync(2)` was not being used with `syncsnoop.bt`.

### CODE REVIEW

**Comment by basil** [comment]
```
15:41:55  HttpMethod: POST
15:41:55  URL: https://incrementals.jenkins.io/
15:41:55  Content-Type: application/json
15:41:55  Authorization: *****
15:41:55  Sending request to url: https://incrementals.jenkins.io/
15:42:55  Response Code: HTTP/1.1 504 Gateway Time-out
15:42:55  Response: 
15:42:55  <html>
15:42:55  <head><title>504 Gateway Time-out</title></head>
15:42:55  <body>
15:42:55  <center><h1>504 Gateway Time-out</h1></center>
15:42:55  <hr><center>nginx</center>
15:42:55  </body>
15:42:55  </html>
```

**Review by bitwiseman — APPROVED** [review]

File: core/src/main/java/hudson/XmlFile.java
Comment (jglick) [inline core/src/main/java/hudson/XmlFile.java:194]: I feel like rather than an arg to an overload of `write`, this would be more natural as a constructor parameter, or a fluent-style method

```java
public XmlFile withoutForce() {
    force = false;
    return this;
}
```

File: core/src/main/java/hudson/XmlFile.java
Comment (jglick) [inline core/src/main/java/hudson/XmlFile.java:202]: https://github.com/jenkinsci/jenkins/blob/23a941f5b44ff5ef7dc5ef7899d4769f7138f0f0/core/src/main/java/hudson/util/AtomicFileWriter.java#L113-L123 is deprecated. But more OK here? https://github.com/jenkinsci/jenkins/pull/3170 @batmat may have an opinion

Comment (basil) [reply core/src/main/java/hudson/XmlFile.java:202]: This seems like an abuse of `@Deprecated`. The method is perfectly OK, but you should think twice about the tradeoffs. I clearly did that.

Comment (jglick) [reply core/src/main/java/hudson/XmlFile.java:205]: Feel free to remove the annotation if you think it is justifiable to call the method under some circumstances. I agree that `@Deprecated` should not be used merely because you think it is important that a caller read and understand the Javadoc. (case in point: `Run.getAbsoluteUrl`)

File: core/src/main/java/hudson/XmlFile.java
Comment (jglick) [inline core/src/main/java/hudson/XmlFile.java:130]: This will not be visible to callers; place Javadoc on the constructor.

File: core/src/main/java/hudson/XmlFile.java
Comment (basil) [reply core/src/main/java/hudson/XmlFile.java:204]: I'll leave that to Baptiste, since he's the one who introduced this. With a logical change that spans five repositories, I'm trying to minimize the amount of unrelated sprawl.

**Review by jglick — APPROVED** [review]

**Comment by basil** [comment]
I addressed all review feedback a week ago, and I just updated all PRs with the latest incrementals today. From my perspective, this is ready for merge.

File: core/src/main/java/hudson/XmlFile.java
Comment (daniel-beck) [inline core/src/main/java/hudson/XmlFile.java:138]: Lacks `@since`

Comment (basil) [reply core/src/main/java/hudson/XmlFile.java:138]: I've addressed your comments in the past and still not gotten approvals from you, so I'm less inclined to address comments like this. Feel free to open a PR if you think this should be fixed.

Comment (daniel-beck) [reply core/src/main/java/hudson/XmlFile.java:138]: It's far easier with limited time to point out weaknesses than approve a nontrivial PR in its entirety.

This one isn't even personal preference (where your response would make sense), if there was a tool for this we'd use it.

Comment (basil) [reply core/src/main/java/hudson/XmlFile.java:138]: Perhaps a bit too easy, especially compared to writing the PR in the first place. By reviewing a PR, you are taking at least partial responsibility for shepherding the PR to completion. While all of us are guilty of so-called "drive-by reviews" from time to time, I think it is a behavior best avoided in healthy code review cultures.

We're getting off topic here. I opened #5661 for the `@since` annotation. Approve it — or not. It's up to you.

Comment (daniel-beck) [reply core/src/main/java/hudson/XmlFile.java:138]: Good point; I'll try to do better.

**Comment by jtnord** [comment]
How I'd this survivable non atomic. You get zero byte XML files by boot using force

File: core/src/main/java/hudson/XmlFile.java
Comment (jtnord) [inline core/src/main/java/hudson/XmlFile.java:138]: this comment is inocreect. 
neither ` FileChannel#force` nor `FlushFileBuffers` will be called before this method returns.
1) this is a constructor not a method
2) there is no data written here.

File: core/src/main/java/hudson/XmlFile.java
Comment (jtnord) [reply core/src/main/java/hudson/XmlFile.java:204]: this method should have been restricted and not exposed (but IIRC it was used so we could not at the time)

people have an expectation of atomicFileWriter - it is clearly documented "It will work, or not, but never have work and corrupt stuff".  so now you hare using an AtomicFileWriter in a way that is clearly not going to do that and it is confusing.

https://github.com/jenkinsci/jenkins/blob/23a941f5b44ff5ef7dc5ef7899d4769f7138f0f0/core/src/main/java/hudson/util/AtomicFileWriter.java#L49-L50

**Comment by basil** [comment]
> people have an expectation

People also have an expectation that on the most commonly used platform (Linux) running a Pipeline job shouldn't call `fsync(2)` hundreds of times per minute.

> How Is this survivable non atomic

It's not. I'm intentionally repurposing the name for a different concept. In the design in [JENKINS-66001](https://issues.jenkins.io/browse/JENKINS-66001), I originally proposed a new name, but Devin wrote:

> Even if the naming is confusing, I would not recommend introducing a new Pipeline durability level with this behavior, I would just update the implementation of `SURVIVABLE_NONATOMIC`, since the proposed behavior is probably how users expect that setting to work anyway.

Note that I did my due diligence by having a design discussion in JENKINS-66001. I formed consensus on the design with Jesse, Devin, and Raihaan. Then I opened this PR. It remained open for a month waiting for review feedback, and I addressed all feedback. If you had feedback to leave, you should have left it then.

> This comment is incorrect.

Indeed. The `force` parameter was originally on the `write` method in an earlier version of this PR. I moved it to the constructor in response to Jesse's review feedback but forgot to update the Javadoc. A trivial mistake.

What is the purpose of leaving a review on a PR that has already been merged? If the PR introduced a regression, that would be one thing, but this PR hasn't caused any regressions. The only flaws have been trivial Javadoc mistakes. If you want to be constructive, how about being part of the solution by opening a PR to fix the trivial Javadoc mistake? Let's work with each other for the benefit of Jenkins users who want better performance.

**Pull request merged**
